### ISSUE TITLE
Equivalence trait instead of Borrow

### ISSUE THREAD

**Issue by stepancheg** [xacrimon/dashmap#183]
See how `IndexMap` is doing it: https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html#method.get

`Equivalent` trait is implemented automatically for `Borrow`, but additionally it can be implemented for complex types. Consider this example:

```
struct Key {
  k0: String,
  k1: String,
}

let map: DashMap<Key, u32> = ...

struct KeyRef<'a> {
  k0: &'a str,
  k1: &'a str,
}

let key_ref: KeyRef = ...

map.get(???)
```

With `indexmap` it is possible to implement `KeyRef: Equivalent<Key>` and query by `KeyRef`.

It is not possible to implement reasonable `Borrow` for `Key` (without allocating).

**Comment by xacrimon** [comment]
Looked into this. It seems to be currently blocked by https://github.com/rust-lang/rust/issues/56167 as we use the std hashmap under the hood.

**Comment by xacrimon** [comment]
Now unblocked as we've switched to hashbrown.

**Comment by tomkarw** [comment]
@xacrimon This one's pretty straightforward. PR and merge?

**Comment by stepancheg** [comment]
Common `Equivalent` now lives in a separate crate by the way: https://docs.rs/equivalent/latest/equivalent/

**Comment by tomkarw** [comment]
> Common Equivalent now lives in a separate crate by the way: https://docs.rs/equivalent/latest/equivalent/

That's true, but we care about whatever `Equivalent` trait hashbrown uses. So I think it's correct that the reexport chain is `equivalent -> hashbrown -> dashmap`.

**Comment by stepancheg** [comment]
> we care about whatever Equivalent trait hashbrown uses

Users of indexmap may prefer to implement `Equivalent` for their types from `equivalent` trait once for all possible map implementations using equivalent, without dependency on hashbrown, or indexmap, or dashmap.

**Comment by stepancheg** [comment]
@tomkarw well, this crate exports hashbrown `HashMap` in public API, then you are correct, it should be hashbrown's `Equivalent`. I just learned it now.

**Comment by xacrimon** [comment]
Hi @tfreiberg-fastly! `master` currently has breaking changesm from v6 and I'm planning to publish a `7.0.0-rc1` next week. Will probably look at a real v7 after a month or so without reported issues. `entry_ref` API would be welcome and merged ASAP. I'll try to get an equivalent API in this weekend for the release candidate.

**Comment by TimoFreiberg** [comment]
alright! i'll try to open a PR tomorrow!
